File: app/schemas/delivery_session.py
from datetime import date
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel
from pydantic import ConfigDict
from pydantic import Field


class DeliverySessionCreate(BaseModel):

    route_id: int

    delivery_date: date

    shift: str = Field(
        pattern="^(MORNING|EVENING)$"
    )

    delivery_partner_id: int


class DeliverySessionUpdate(BaseModel):

    status: str | None = Field(
        default=None,
        pattern="^(PLANNED|STARTED|COMPLETED|CLOSED)$"
    )

    total_milk_loaded: Decimal | None = Field(
        default=None,
        ge=0
    )

    total_cash_sales: Decimal | None = Field(
        default=None,
        ge=0
    )

    total_returned_milk: Decimal | None = Field(
        default=None,
        ge=0
    )

    reconciliation_status: str | None = Field(
        default=None,
        pattern="^(BALANCED|UNBALANCED|PENDING)$"
    )


class DeliverySessionDispatch(BaseModel):

    total_milk_loaded: Decimal = Field(
        gt=0
    )


class DeliverySessionReopen(BaseModel):

    reason: str = Field(
        min_length=1,
        max_length=500
    )


class DailyDeliveryResponse(BaseModel):

    id: int

    customer_id: int

    customer_name: str | None = None

    milk_type_name: str | None = None

    planned_quantity: int

    delivered_quantity: int

    delivery_status: str

    delivery_source: str

    token_sheet_number: int | None = None

    model_config = ConfigDict(
        from_attributes=True
    )


class DeliverySessionResponse(BaseModel):

    id: int

    route_id: int

    route_name: str | None = None

    delivery_date: date

    shift: str

    delivery_partner_id: int

    delivery_partner_name: str | None = None

    status: str

    total_milk_loaded: Decimal

    total_token_registered: Decimal

    total_cash_sales: Decimal

    total_returned_milk: Decimal

    reconciliation_status: str

    reopen_count: int

    version: int

    created_at: datetime

    updated_at: datetime

    model_config = ConfigDict(
        from_attributes=True
    )


class DeliverySessionDetailResponse(DeliverySessionResponse):

    deliveries: list[DailyDeliveryResponse] = []


class DeliverySessionListResponse(BaseModel):

    sessions: list[DeliverySessionResponse]

    total: int


class ChecklistCustomer(BaseModel):

    customer_id: int

    customer_name: str

    address: str | None = None

    phone: str | None = None

    milk_type: str

    quantity: int

    model_config = ConfigDict(
        from_attributes=True
    )


class DeliveryChecklistResponse(BaseModel):

    session_id: int

    route_name: str | None = None

    delivery_date: date

    shift: str

    total_expected: int

    customers: list[ChecklistCustomer]


class ReconciliationResponse(BaseModel):

    session_id: int

    loaded_milk: Decimal

    token_registered: Decimal

    cash_sales: Decimal

    returned_milk: Decimal

    total_accounted: Decimal

    difference: Decimal

    is_balanced: bool

    status: str


class SessionReportSummary(BaseModel):

    total_customers: int

    delivered: int

    pending_token: int

    cash_sale: int

    not_delivered: int


class SessionReportMilkSummary(BaseModel):

    loaded: Decimal

    token_registered: Decimal

    cash_sales: Decimal

    returned: Decimal


class SessionReportResponse(BaseModel):

    session_id: int

    route_name: str | None = None

    delivery_date: date

    shift: str

    summary: SessionReportSummary

    milk_summary: SessionReportMilkSummary

File: frontend/src/types/delivery-session.ts
export type DeliverySessionStatus = "PLANNED" | "STARTED" | "COMPLETED" | "CLOSED";

export type Shift = "MORNING" | "EVENING";

export type ReconciliationStatus = "BALANCED" | "UNBALANCED" | "PENDING";

export interface DeliverySessionCreate {
  route_id: number;
  delivery_date: string;
  shift: Shift;
  delivery_partner_id: number;
}

export interface DeliverySessionDispatch {
  total_milk_loaded: number;
}

export interface DeliverySessionReopen {
  reason: string;
}

export interface DeliverySessionResponse {
  id: number;
  route_id: number;
  route_name: string | null;
  delivery_date: string;
  shift: Shift;
  delivery_partner_id: number;
  delivery_partner_name: string | null;
  status: DeliverySessionStatus;
  total_milk_loaded: number;
  total_token_registered: number;
  total_cash_sales: number;
  total_returned_milk: number;
  reconciliation_status: ReconciliationStatus;
  reopen_count: number;
  version: number;
  created_at: string;
  updated_at: string;
}

export interface DeliverySessionDeliverySummary {
  id: number;
  customer_id: number;
  customer_name: string | null;
  milk_type_name: string | null;
  planned_quantity: number;
  delivered_quantity: number;
  delivery_status: string;
  delivery_source: string;
  token_sheet_number: number | null;
}

export interface DeliverySessionDetailResponse extends DeliverySessionResponse {
  deliveries: DeliverySessionDeliverySummary[];
}

export interface DeliverySessionListResponse {
  sessions: DeliverySessionResponse[];
  total: number;
}

export interface ReconciliationResponse {
  session_id: number;
  loaded_milk: number;
  token_registered: number;
  cash_sales: number;
  returned_milk: number;
  total_accounted: number;
  difference: number;
  is_balanced: boolean;
  status: ReconciliationStatus;
}

export interface ReconciliationValidationIssue {
  code: string;
  message: string;
  severity: string;
}

export interface ReconciliationValidationResponse {
  can_close: boolean;
  is_balanced: boolean;
  issues: ReconciliationValidationIssue[];
}

export interface SessionReportSummary {
  total_customers: number;
  delivered: number;
  pending_token: number;
  cash_sale: number;
  not_delivered: number;
}

export interface SessionReportMilkSummary {
  loaded: number;
  token_registered: number;
  cash_sales: number;
  returned: number;
}

export interface SessionReportResponse {
  session_id: number;
  route_name: string | null;
  delivery_date: string;
  shift: Shift;
  summary: SessionReportSummary;
  milk_summary: SessionReportMilkSummary;
}
